com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs

```diff
@@ -202,22 +202,33 @@ void AddItemRequested(Blackboard blackboard)
 
         void AddPropertyItems(GenericMenu gm)
         {
-            gm.AddItem(new GUIContent($"Vector1"), false, () => AddInputRow(new Vector1ShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Vector2"), false, () => AddInputRow(new Vector2ShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Vector3"), false, () => AddInputRow(new Vector3ShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Vector4"), false, () => AddInputRow(new Vector4ShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Color"), false, () => AddInputRow(new ColorShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Texture2D"), false, () => AddInputRow(new Texture2DShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Texture2D Array"), false, () => AddInputRow(new Texture2DArrayShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Texture3D"), false, () => AddInputRow(new Texture3DShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Cubemap"), false, () => AddInputRow(new CubemapShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Virtual Texture"), false, () => AddInputRow(new VirtualTextureShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Boolean"), false, () => AddInputRow(new BooleanShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Matrix2x2"), false, () => AddInputRow(new Matrix2ShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Matrix3x3"), false, () => AddInputRow(new Matrix3ShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Matrix4x4"), false, () => AddInputRow(new Matrix4ShaderProperty(), true));
-            gm.AddItem(new GUIContent($"SamplerState"), false, () => AddInputRow(new SamplerStateShaderProperty(), true));
-            gm.AddItem(new GUIContent($"Gradient"), false, () => AddInputRow(new GradientShaderProperty(), true));
+            Dictionary<Type, BlackboardInputInfo> blackboardInputInfos = new Dictionary<Type, BlackboardInputInfo>();
+
+            var shaderInputTypes = TypeCache.GetTypesDerivedFrom<ShaderInput>();
+
+            // Sort the ShaderInput by priority using the BlackboardInputInfo attribute
+            var sortedInputTypes = shaderInputTypes.OrderBy(t => {
```

Comment:
May be worth using Sort() so you can sort by priority first, then alphabetical for non-priority items.  Would also let us remove most of the priority attribute tags if we don't care about their ordering.

Reply:
I thought Sort was a bit too much, but if you want I can do something like this:

```CSharp
            shaderInputTypes.Sort((s1, s2) => {
                var info1 = Attribute.GetCustomAttribute(t, typeof(BlackboardInputInfo)) as BlackboardInputInfo;
                var info2 = Attribute.GetCustomAttribute(t, typeof(BlackboardInputInfo)) as BlackboardInputInfo;
            
                if (info1 == null && info2 == null)
                    return s1.Name.CompareTo(s2.Name);
                else if (info1 != null && info2 == null)
                    return 1;
                else if (info1 == null && info2 != null)
                    return -1;
                else // Both info1 and info2 are not null
                {
                    if (info1.priority == info2.priority)
                        return (info1.name ?? s1.Name).CompareTo(info2.name ?? s2.Name);
                    else
                        return info1.priority.CompareTo(info2.priority);
                }
            });
```

Reply:
yeah that looks great!

Reply:
I noticed we have an issue with another PR in flight here (https://github.com/Unity-Technologies/Graphics/pull/381) -- where the Texture2DArrayShaderProperty was subclassed to define special behavior for Lightmap properties..  but we don't want this property type to show up in the menu.

I was thinking: if we make the BlackboardInputInfo attribute required to show up in the menu here, that would let us easily hide internal types by just not tagging them.

And we could use to filter out keywords as well.

Reply:
done